File: src/shared/relay-transport/HostStream.ts
/** Controller-side reader of one host-owned stream.
 *
 * Wire contract shared with `remote_connect/host_stream.rs`: every page is read
 * from the online host over pairwise-encrypted device RPC (`read_stream`), and
 * the host pushes `host-stream-changed` hints naming only the stream, its
 * epoch and newest sequence. The Relay forwards ciphertext and stores nothing;
 * this client keeps no cache either, so closing it forgets the transcript. */

export const HOST_CATALOG_ID = '@host/catalog';
export const HOST_STREAM_CHANGED_EVENT = 'host-stream-changed';
export const REMOTE_CAPABILITY_HOST_STREAM_V1 = 'host_stream_v1';
export const UNSUPPORTED_HOST_MESSAGE = 'The controlled device runs an older BitFun version that does not support on-demand session streaming. Update BitFun on that device to continue.';
/** Hosts keep hint subscriptions alive for 10 minutes; renew well before. */
const KEEPALIVE_MS = 4 * 60 * 1000;

export interface StreamEvent { seq: number; event: string; payload: unknown }
export interface StreamPage {
  stream_id: string;
  epoch: number;
  events: StreamEvent[];
  has_more: boolean;
  cursor: number;
  oldest_seq: number;
  truncated: boolean;
}
export interface StreamReadRequest {
  stream_id: string;
  after?: number;
  before?: number;
  epoch?: number;
  limit?: number;
  subscribe: boolean;
}
export interface StreamHint { sourceDeviceId: string; stream_id: string; epoch: number; cursor: number }

/** Event as consumed by product reducers; `session_id` is the stream id. */
export interface SessionEvent { session_id: string; event: string; payload: unknown }
export interface SessionHistoryState { hasMore: boolean; oldestSeq: number; cursor: number; truncated: boolean }
export interface SessionStreamHandle { close(): void; wake(): void; loadOlder(): Promise<void> }

/** Host-facing calls of one reader; production is encrypted device RPC. */
export interface HostStreamTransport {
  readStream(request: StreamReadRequest): Promise<StreamPage>;
  unsubscribeStream(streamId: string): Promise<void>;
}
/** Wake sources: decrypted hints and connection recoveries. */
export interface HostStreamSignals {
  onHint(listener: (hint: StreamHint) => void): () => void;
  onReconnect(listener: () => void): () => void;
}

export interface HostStreamOptions {
  transport: HostStreamTransport;
  signals: HostStreamSignals;
  /** Device id of the host whose hints apply to this reader. */
  target: string;
  streamId: string;
  onEvent: (event: SessionEvent) => void;
  onError: (error: unknown) => void;
  /** Called after every completed catch-up, including the opening page. */
  onCaughtUp?: () => void;
  onHistoryState?: (state: SessionHistoryState) => void;
  /** Called on every wake that re-reads the host (hint, reconnect, visibility). */
  onResumed?: () => void;
  /** The host restarted the stream; consumers must drop derived state before
   * the latest page is replayed through `onEvent`. */
  onGap?: (reason: string) => void;
  /** Page visibility owner; defaults to `document` when present. */
  visibility?: Pick<Document, 'addEventListener' | 'removeEventListener' | 'visibilityState'> | null;
}

/** Map an older host's "unknown command" rejection to an actionable message. */
export function describeHostStreamError(error: unknown): string {
  const text = error instanceof Error ? error.message : String(error);
  if (text.includes('invalid RPC command') || text.includes('unknown variant') || text.includes('Could not parse device command')) {
    return UNSUPPORTED_HOST_MESSAGE;
  }
  return text;
}

export function parseStreamHint(sourceDeviceId: string, event: string, payload: unknown): StreamHint | null {
  if (event !== HOST_STREAM_CHANGED_EVENT || !payload || typeof payload !== 'object') return null;
  const hint = payload as { stream_id?: unknown; epoch?: unknown; cursor?: unknown };
  if (typeof hint.stream_id !== 'string' || typeof hint.epoch !== 'number' || typeof hint.cursor !== 'number') return null;
  return { sourceDeviceId, stream_id: hint.stream_id, epoch: hint.epoch, cursor: hint.cursor };
}

export function parseStreamPage(value: unknown): StreamPage {
  const page = value as Partial<StreamPage> & { resp?: unknown; message?: unknown };
  if (!page || typeof page !== 'object') throw new Error('Invalid stream page');
  if (page.resp === 'error') throw new Error(typeof page.message === 'string' ? page.message : 'Remote error');
  if (page.resp !== undefined && page.resp !== 'stream_page') throw new Error(`Unexpected stream response: ${String(page.resp)}`);
  if (typeof page.stream_id !== 'string' || !Number.isSafeInteger(page.epoch) || !Array.isArray(page.events)
    || typeof page.has_more !== 'boolean' || !Number.isSafeInteger(page.cursor) || !Number.isSafeInteger(page.oldest_seq)) {
    throw new Error('Invalid stream page');
  }
  for (const event of page.events) {
    const item = event as Partial<StreamEvent>;
    if (!item || typeof item !== 'object' || !Number.isSafeInteger(item.seq) || typeof item.event !== 'string' || !('payload' in item)) {
      throw new Error('Invalid stream event');
    }
  }
  return {
    stream_id: page.stream_id, epoch: page.epoch!, events: page.events as StreamEvent[], has_more: page.has_more!,
    cursor: page.cursor!, oldest_seq: page.oldest_seq!, truncated: page.truncated === true,
  };
}

class Reader {
  epoch = 0;
  cursor = 0;
  oldest = 1;
  hasMore = false;
  truncated = false;
  constructor(private readonly options: HostStreamOptions) {}

  private async read(request: Omit<StreamReadRequest, 'stream_id' | 'subscribe'>): Promise<StreamPage> {
    const page = await this.options.transport.readStream({ stream_id: this.options.streamId, subscribe: true, ...request });
    if (page.stream_id !== this.options.streamId) throw new Error('Stream page identity mismatch');
    return page;
  }
  private emitPage(page: StreamPage): void {
    for (const event of page.events) {
      this.options.onEvent({ session_id: this.options.streamId, event: event.event, payload: event.payload });
    }
  }
  private emitHistory(): void {
    this.options.onHistoryState?.({ hasMore: this.hasMore, oldestSeq: this.oldest, cursor: this.cursor, truncated: this.truncated });
  }
  /** Latest page first, like opening a chat at its bottom. */
  async resync(): Promise<void> {
    const page = await this.read({});
    this.epoch = page.epoch;
    this.cursor = page.cursor;
    this.oldest = page.events[0]?.seq ?? page.cursor + 1;
    this.hasMore = page.has_more;
    this.truncated = page.truncated;
    this.emitPage(page);
    this.emitHistory();
  }
  /** False when the host restarted the stream and a resync is required. */
  private async catchUp(): Promise<boolean> {
    for (;;) {
      const page = await this.read({ after: this.cursor, epoch: this.epoch });
      if (page.epoch !== this.epoch) return false;
      this.emitPage(page);
      const last = page.events[page.events.length - 1];
      if (last) this.cursor = Math.max(this.cursor, last.seq);
      if (!page.has_more) {
        // The newest sequence may belong to an evicted control event; adopt
        // it so the next hint compares against the host's view.
        this.cursor = Math.max(this.cursor, page.cursor);
        return true;
      }
    }
  }
  async refresh(): Promise<void> {
    if (await this.catchUp()) return;
    this.options.onGap?.('host stream restarted');
    await this.resync();
  }
  async loadOlder(): Promise<void> {
    if (!this.hasMore) return;
    const page = await this.read({ before: this.oldest, epoch: this.epoch });
    if (page.epoch !== this.epoch) {
      this.options.onGap?.('host stream restarted');
      await this.resync();
      throw new Error('Session history restarted on the host; reloaded from its latest page');
    }
    this.emitPage(page);
    if (page.events[0]) this.oldest = page.events[0].seq;
    this.hasMore = page.has_more;
    this.truncated = page.truncated;
    this.emitHistory();
  }
  hintIsNew(hint: StreamHint): boolean {
    return hint.sourceDeviceId === this.options.target && hint.stream_id === this.options.streamId
      && (hint.epoch !== this.epoch || hint.cursor > this.cursor);
  }
}

/** Open one host stream. The first page is awaited so an offline or older
 * host fails the call instead of retrying silently in the background. */
export async function openHostStream(options: HostStreamOptions): Promise<SessionStreamHandle> {
  const reader = new Reader(options);
  let active = true;
  try { await reader.resync(); } catch (error) { throw new Error(describeHostStreamError(error)); }
  options.onCaughtUp?.();

  // One reader in flight and one dirty successor (Happy InvalidateSync);
  // older-page requests share the same lane so pages never interleave.
  let dirty = false;
  let running: Promise<void> | null = null;
  let retry: ReturnType<typeof setTimeout> | null = null;
  let retryDelay = 1000;
  const olderRequests: Array<{ resolve: () => void; reject: (error: unknown) => void }> = [];
  const drain = async () => {
    while (active && (dirty || olderRequests.length)) {
      if (dirty) {
        dirty = false;
        await reader.refresh();
        if (active) options.onCaughtUp?.();
      }
      const older = olderRequests.shift();
      if (!older) continue;
      try { await reader.loadOlder(); older.resolve(); } catch (error) { older.reject(error); }
    }
  };
  const wake = () => {
    if (!active || running || retry) return;
    running = drain().then(() => { retryDelay = 1000; }).catch(error => {
      if (!active) return;
      dirty = true;
      retry = setTimeout(() => { retry = null; wake(); }, retryDelay);
      retryDelay = Math.min(30000, retryDelay * 2);
      options.onError(new Error(describeHostStreamError(error)));
    }).finally(() => {
      running = null;
      if (active && (dirty || olderRequests.length) && !retry) wake();
    });
  };
  const invalidate = () => { if (!active) return; dirty = true; wake(); };
  const resumed = () => { if (!active) return; options.onResumed?.(); invalidate(); };

  const stopHints = options.signals.onHint(hint => { if (active && reader.hintIsNew(hint)) invalidate(); });
  const stopReconnect = options.signals.onReconnect(resumed);
  const keepalive = setInterval(invalidate, KEEPALIVE_MS);
  const visibility = options.visibility === undefined ? (typeof document === 'undefined' ? null : document) : options.visibility;
  const onVisible = () => { if (visibility?.visibilityState === 'visible') resumed(); };
  visibility?.addEventListener('visibilitychange', onVisible);

  return {
    close() {
      if (!active) return;
      active = false;
      stopHints(); stopReconnect(); clearInterval(keepalive);
      visibility?.removeEventListener('visibilitychange', onVisible);
      if (retry) clearTimeout(retry);
      for (const older of olderRequests.splice(0)) older.reject(new Error('Session stream closed'));
      void options.transport.unsubscribeStream(options.streamId).catch(() => {});
    },
    wake: invalidate,
    loadOlder() {
      if (!active) return Promise.resolve();
      return new Promise<void>((resolve, reject) => { olderRequests.push({ resolve, reject }); wake(); });
    },
  };
}

File: src/apps/mobile/shared/core-transport/src/commonMain/kotlin/com/bitfun/mobile/core/transport/HostStream.kt
package com.bitfun.mobile.core.transport

import com.bitfun.mobile.core.protocol.CommandStatus
import com.bitfun.mobile.core.protocol.isError
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/*
 * Controller-side reader of one host-owned stream.
 *
 * Wire contract shared with `remote_connect/host_stream.rs` and the Web
 * `HostStream.ts`: every page is read from the online desktop over pairwise
 * encrypted device RPC (`read_stream`), and the desktop pushes
 * `host-stream-changed` hints naming only the stream, its epoch and newest
 * sequence. The relay forwards ciphertext and stores nothing; this client keeps
 * no cache either, so closing the flow forgets the transcript.
 */

public const val HOST_CATALOG_ID: String = "@host/catalog"
public const val HOST_STREAM_CHANGED_EVENT: String = "host-stream-changed"
public const val REMOTE_CAPABILITY_HOST_STREAM_V1: String = "host_stream_v1"

/** Hosts keep hint subscriptions alive for 10 minutes; renew well before. */
internal const val HOST_STREAM_KEEPALIVE_MS: Long = 4 * 60 * 1000L

/** Pseudo events the stream emits around the host's own records. */
public const val STREAM_EVENT_READY: String = "relay://session-ready"
public const val STREAM_EVENT_RESUMED: String = "relay://session-resumed"
public const val STREAM_EVENT_GAP: String = "relay://session-gap"

/**
 * The controlled device predates on-demand streams. Nothing on this side can
 * make it answer `read_stream`; the user has to update that device.
 */
public class HostStreamUnsupportedException public constructor(cause: Throwable?) :
    RelayTransportException(RelayFailure.HostStreamUnsupported, cause) {
    public constructor() : this(null)
}

@Serializable
internal data class StreamEventWire(val seq: Long, val event: String, val payload: JsonElement = JsonNull)

/** `RemoteResponse::StreamPage` as flattened onto the wire. */
@Serializable
internal data class StreamPageWire(
    @SerialName("resp") override val resp: String? = null,
    @SerialName("message") override val message: String? = null,
    @SerialName("stream_id") val streamId: String = "",
    val epoch: Long = -1,
    val events: List<StreamEventWire> = emptyList(),
    @SerialName("has_more") val hasMore: Boolean = false,
    val cursor: Long = -1,
    @SerialName("oldest_seq") val oldestSeq: Long = -1,
    val truncated: Boolean = false,
) : CommandStatus

/** A decrypted `host-stream-changed` device event. */
public data class StreamHint(
    public val sourceDeviceId: String,
    public val streamId: String,
    public val epoch: Long,
    public val cursor: Long,
)

/** Parses a decrypted `device_event` plaintext; null for anything that is not a stream hint. */
internal fun parseStreamHint(sourceDeviceId: String, plaintext: JsonObject): StreamHint? {
    if ((plaintext["cmd"] as? JsonPrimitive)?.content != "device_event") return null
    if ((plaintext["event"] as? JsonPrimitive)?.content != HOST_STREAM_CHANGED_EVENT) return null
    val payload = plaintext["payload"] as? JsonObject ?: return null
    val streamId = (payload["stream_id"] as? JsonPrimitive)?.takeIf { it.isString }?.content ?: return null
    val epoch = (payload["epoch"] as? JsonPrimitive)?.takeIf { !it.isString }?.content?.toLongOrNull() ?: return null
    val cursor = (payload["cursor"] as? JsonPrimitive)?.takeIf { !it.isString }?.content?.toLongOrNull() ?: return null
    return StreamHint(sourceDeviceId, streamId, epoch, cursor)
}

/**
 * Whether a desktop rejection means "this host has no `read_stream`", as
 * opposed to a refusal it chose. Older hosts answer an unknown `cmd` with a
 * parse failure; there is no capability bit to consult on that path.
 */
internal fun isUnknownStreamCommandRejection(message: String?): Boolean {
    val text = message.orEmpty()
    return text.contains("invalid RPC command") || text.contains("unknown variant") ||
        text.contains("Could not parse device command")
}

internal fun isRetryableStreamFailure(failure: Throwable): Boolean =
    failure is CloudAccountException && failure.failure in setOf(
        CloudAccountFailure.NETWORK, CloudAccountFailure.TIMEOUT,
        CloudAccountFailure.RELAY_UNAVAILABLE, CloudAccountFailure.RATE_LIMITED,
    )

/** Rejects a page the host did not accept and asserts it answers this stream. */
internal fun checkStreamPage(streamId: String, page: StreamPageWire): StreamPageWire {
    if (page.isError) {
        if (isUnknownStreamCommandRejection(page.message)) throw HostStreamUnsupportedException()
        throw RelayTransportException(RelayFailure.RemoteRejected(page.message))
    }
    if (page.resp != null && page.resp != "stream_page") throw CloudAccountException(CloudAccountFailure.MALFORMED_RESPONSE)
    if (page.streamId != streamId || page.epoch < 0 || page.cursor < 0) throw CloudAccountException(CloudAccountFailure.MALFORMED_RESPONSE)
    return page
}

/** Host reads issued by one stream; production is encrypted device RPC. */
internal interface HostStreamReads {
    suspend fun read(after: Long?, before: Long?, epoch: Long?): StreamPageWire
    suspend fun unsubscribe()
}

private class HostStreamReader(
    private val streamId: String,
    private val reads: HostStreamReads,
    private val emit: suspend (JsonObject) -> Unit,
) {
    var epoch = 0L
    var cursor = 0L
    var oldest = 1L
    var hasMore = false
    var truncated = false

    private suspend fun read(after: Long? = null, before: Long? = null, epoch: Long? = null): StreamPageWire =
        checkStreamPage(streamId, reads.read(after, before, epoch))

    private suspend fun emitPage(page: StreamPageWire) {
        for (event in page.events) {
            emit(buildJsonObject { put("session_id", streamId); put("event", event.event); put("payload", event.payload) })
        }
    }

    private suspend fun emitReady() {
        emit(buildJsonObject {
            put("session_id", streamId); put("event", STREAM_EVENT_READY)
            put("payload", buildJsonObject { put("hasMore", hasMore); put("oldestSeq", oldest); put("cursor", cursor); put("truncated", truncated) })
        })
    }

    private suspend fun emitGap() {
        emit(buildJsonObject {
            put("session_id", streamId); put("event", STREAM_EVENT_GAP)
            put("payload", buildJsonObject { put("reason", "host stream restarted") })
        })
    }

    /** Latest page first, like opening a chat at its bottom. */
    suspend fun resync() {
        val page = read()
        epoch = page.epoch
        cursor = page.cursor
        oldest = page.events.firstOrNull()?.seq ?: (page.cursor + 1)
        hasMore = page.hasMore
        truncated = page.truncated
        emitPage(page)
        emitReady()
    }

    /** False when the host restarted the stream and a resync is required. */
    private suspend fun catchUp(): Boolean {
        while (true) {
            val page = read(after = cursor, epoch = epoch)
            if (page.epoch != epoch) return false
            emitPage(page)
            page.events.lastOrNull()?.let { cursor = maxOf(cursor, it.seq) }
            if (!page.hasMore) {
                // The newest sequence may belong to an evicted control event;
                // adopt it so the next hint compares against the host's view.
                cursor = maxOf(cursor, page.cursor)
                return true
            }
        }
    }

    suspend fun refresh() {
        if (catchUp()) return
        emitGap()
        resync()
    }

    suspend fun loadOlder() {
        if (!hasMore) return
        val page = read(before = oldest, epoch = epoch)
        if (page.epoch != epoch) {
            emitGap()
            resync()
            throw IllegalStateException("Session history restarted on the host; reloaded from its latest page")
        }
        emitPage(page)
        page.events.firstOrNull()?.let { oldest = it.seq }
        hasMore = page.hasMore
        truncated = page.truncated
        emitReady()
    }

    fun hintIsNew(hint: StreamHint, target: String): Boolean =
        hint.sourceDeviceId == target && hint.streamId == streamId && (hint.epoch != epoch || hint.cursor > cursor)
}

private sealed interface Wake {
    /** Re-read the host; [announce] marks a connection or foreground recovery. */
    class Dirty(val announce: Boolean) : Wake
    class Older(val request: CompletableDeferred<Unit>) : Wake
}

/**
 * One host stream as a flow of `{session_id, event, payload}` objects.
 *
 * The opening page is read before the first emission, retrying transient
 * relay failures with backoff while [onError] reports each one; a host that
 * refuses or cannot parse `read_stream` fails the flow instead. Afterwards
 * hints, reconnects and the keepalive share one reader lane with history
 * requests so pages never interleave.
 *
 * Every wake that re-reads the host after a recovery emits
 * [STREAM_EVENT_RESUMED] first; a host restart emits [STREAM_EVENT_GAP] before
 * the latest page is replayed, and each completed catch-up calls [onCaughtUp].
 */
internal fun hostStream(
    streamId: String,
    target: String,
    hints: Flow<StreamHint>,
    reconnects: Flow<Long>,
    reads: HostStreamReads,
    olderRequests: Channel<CompletableDeferred<Unit>>,
    onError: (Throwable) -> Unit,
    onCaughtUp: () -> Unit,
    keepaliveMs: Long = HOST_STREAM_KEEPALIVE_MS,
): Flow<JsonObject> = channelFlow {
    val reader = HostStreamReader(streamId, reads) { send(it) }
    // Set once a page was read with `subscribe`; before that there is nothing
    // on the host to release.
    var subscribed = false
    try {
        var retryMs = 1000L
        while (true) {
            try {
                reader.resync()
                subscribed = true
                break
            } catch (cancelled: CancellationException) {
                throw cancelled
            } catch (error: Throwable) {
                if (!isRetryableStreamFailure(error)) throw error
                onError(error)
                delay(retryMs)
                retryMs = (retryMs * 2).coerceAtMost(30_000L)
            }
        }
        onCaughtUp()

        val wakes = Channel<Wake>(Channel.UNLIMITED)
        launch { hints.collect { hint -> if (reader.hintIsNew(hint, target)) wakes.send(Wake.Dirty(false)) } }
        launch { reconnects.collect { wakes.send(Wake.Dirty(true)) } }
        launch { while (currentCoroutineContext().isActive) { delay(keepaliveMs); wakes.send(Wake.Dirty(false)) } }
        launch { for (request in olderRequests) wakes.send(Wake.Older(request)) }

        var backoffMs = 1000L
        var dirty = false
        var announce = false
        while (currentCoroutineContext().isActive) {
            var older: CompletableDeferred<Unit>? = null
            when (val wake = wakes.receive()) {
                is Wake.Dirty -> { dirty = true; announce = announce || wake.announce }
                is Wake.Older -> older = wake.request
            }
            try {
                if (dirty) {
                    if (announce) {
                        send(buildJsonObject { put("session_id", streamId); put("event", STREAM_EVENT_RESUMED); put("payload", JsonObject(emptyMap())) })
                    }
                    reader.refresh()
                    dirty = false
                    announce = false
                    onCaughtUp()
                }
                older?.let { request ->
                    try { reader.loadOlder(); request.complete(Unit) } catch (cancelled: CancellationException) { throw cancelled } catch (error: Throwable) { request.completeExceptionally(error) }
                }
                backoffMs = 1000L
            } catch (cancelled: CancellationException) {
                older?.cancel()
                throw cancelled
            } catch (error: Throwable) {
                older?.completeExceptionally(error)
                if (error is HostStreamUnsupportedException) throw error
                onError(error)
                dirty = true
                // The retry is its own wake so a hint arriving meanwhile is not lost.
                val wait = backoffMs
                launch { delay(wait); wakes.send(Wake.Dirty(false)) }
                backoffMs = (backoffMs * 2).coerceAtMost(30_000L)
            }
        }
    } finally {
        // Best effort: the host also drops idle subscriptions on its own.
        if (subscribed) withContext(NonCancellable) {
            try { withTimeout(5_000L) { reads.unsubscribe() } } catch (_: Throwable) {}
        }
    }
}

/** Streams that a command transport can also read on demand from its device. */
public interface RemoteSessionStreamTransport {
    /** Re-read every open stream now, e.g. when the app returns to the foreground. */
    public fun wakeSessionStreams() {}
    public suspend fun loadOlder(sessionId: String) { error("History stream is unavailable") }

    /**
     * Opens one host stream. The returned flow emits the host's own records as
     * `{session_id, event, payload}` plus the `relay://` pseudo events declared
     * above. Nothing is persisted on this device.
     */
    public suspend fun subscribe(sessionId: String, onError: (Throwable) -> Unit, onCaughtUp: () -> Unit): Flow<JsonObject>
}
